libs/core/src/main/java/org/elasticsearch/core/internal/provider/ProviderLocator.java

```diff
@@ -10,34 +10,82 @@
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
+import java.lang.module.Configuration;
+import java.lang.module.ModuleFinder;
 import java.security.AccessController;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
 import java.util.Objects;
+import java.util.ServiceConfigurationError;
 import java.util.ServiceLoader;
+import java.util.Set;
 import java.util.function.Supplier;
 
 /**
  * A provider locator that finds the implementation of the specified provider.
  *
- * <p> A provider locator is given a small recipe, in the form of constructor arguments, which it uses to find the required provider
- * implementation.
+ * <p> A provider locator is given a small recipe, in the form of constructor arguments, which it
+ * uses to find the required provider implementation.
  *
- * <p> When run as a module, the locator will load the provider implementation as a module, in its own module layer.
- * Otherwise, the provider implementation will be loaded as a non-module.
+ * <p> When run as a module, the locator will load the provider implementation as a module, in its
+ * own module layer. Otherwise, the provider implementation will be loaded as a non-module.
  *
  * @param <T> the provider type
  */
 public final class ProviderLocator<T> implements Supplier<T> {
 
     private final String providerName;
     private final Class<T> providerType;
+    private final String providerModuleName;
 
-    public ProviderLocator(String providerName, Class<T> providerType) {
+    private final ClassLoader parentLoader;
+
+    private final Set<String> missingModules;
+
+    // whether to load the provider implementation as a module or not
+    private final boolean loadAsProviderModule;
+
+    public ProviderLocator(
+        Module caller,
+        String providerName,
+        Class<T> providerType,
+        String providerModuleName,
+        Set<String> missingModules
+    ) {
+        this(
+            caller,
+            providerName,
+            providerType,
+            ProviderLocator.class.getClassLoader(),
+            providerModuleName,
+            missingModules,
+            ProviderLocator.class.getModule().isNamed()
+        );
+    }
+
+    public ProviderLocator(
+        Module caller,
```

Comment:
Can we not find this ourselves? Or when would it differ from the module of the providerType?

Reply:
we could, but it would require a stack walk. I could go either way on this, since it's only used for a module declaration check anyway.

Reply:
the check has been simplified and reworked